File: LeetCode/322. 零钱兑换.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        coins.sort()
        self.dfs(amount, 0, coins, 0)

    def dfs(self, amount, sum, coins, times):
        if sum == amount: return times
        odd = self.binary_search(coins, amount-sum)
        if odd == -1: return
        while odd >= 0:
            self.dfs(amount, sum + coins[odd])
            odd = odd - 1

    # 二分查找最后一个小于等于target的值
    def binary_search(self, coins, targer):
        low = 0
        high = len(coins)-1

        while low <= high:
            mid = low + int((high-low) / 2)
            if coins[mid] > targer: high = mid - 1
            elif: low = mid + 1
        return low-1

File: Week_04/322. 零钱兑换.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        Arrays.sort(coins);
        int[] ans = new int[1];
        ans[0] = 1<<31 - 1;
        dfs(amount, coins, coins.length-1, 0, ans);
        return ans[0] == 1<<31 -1?-1:ans[0];
    }
    public void dfs(int amount,int[] coins, int c_index, int times, int[] ans) {
        if (amount == 0) {
            ans[0] = Math.min(ans[0], times);
        }
        if (c_index < 0) return;
        int k = amount / coins[c_index];
        for (int i = k;i >= 0 && k + times < ans[0] ;i --) {
            dfs(amount-coins[c_index]*i, coins, c_index-1, times+i, ans);
        }
    }
}